File: matlab/table_3.m
% Table 3 in Section 8.2.2 in Canay, Illanes and Velez (2023)

% Inputs

% ../data/                               fake data
%  - A.csv               n  x (1+J0)     matrix of revenue differential
%  - D.csv               n  x (1+J)      matrix of all product portfolio
%  - J0.csv              J0 x 2          matrix of ownership by two firms

% compute_An_vec                         find r_n^INF and compute A_n^* defined in Section 4 in Andrews and Kwon (2023)
%  - std_R_vec                           compute sd_1, sd_2, and sd_3 in eq. (4.19), (4.21), and (4.22) in Andrews and Kwon (2023)
%  - An_star                             computes the objective function that appears in inf problem defined in eq. (4.25) in Andrews and Kwon (2023)

% G_restriction.m                        find test statistic and c. value
%  - m_function                          compute (26)-(27)
%  - m_hat                               compute a version of (38)
%  - cvalue_SN2S                         compute c. value as in (41)
%  - cvalue_EB2S                         compute c. value as in (48)
%  - cvalue_SPUR1                        compute c. value as in Section C

% output

% '_results/tables-tex/table_3.tex'       confidence intervals and comp.time

% comment:
% the first column of A_matrix and D_matrix were used to index the markets,
% these first columns are useless in the rest of the code.

clc; clear all; close all;
addpath('1_functions');
mkdir('_results');

%% 1 Setup

% Import data
dgp = struct;
dgp.A_matrix = csvread(fullfile('..', 'data', 'A.csv'));
dgp.D_matrix = csvread(fullfile('..', 'data', 'D.csv'));
dgp.J0_vec = csvread(fullfile('..', 'data', 'J0.csv')); % (product, firm), where firm = 1 coca-cola, firm=2 energy-products

dgp.num_market = size(dgp.A_matrix, 1);
dgp.num_product = size(dgp.D_matrix, 1) - 1;
dgp.W_data = dgp.D_matrix(:, 2:end);

% Settings (cell arrays are used to loop over each of the four different specifications)
settings = struct;
settings.Vbar = {0, 0, 0, 0}; % Vbar is defined in Assumption 4.2 and appears in eq. (26)-(27).
settings.test_stat = {'CCK', 'RC-CCK', 'RC-CCK', 'RC-CCK'}; % CCK as in eq. (38) and RC-CCK is its recentered version as in Section 8.2.2.
settings.cv = {'SN2S', 'SN2S', 'EB2S', 'SPUR1'}; % Critical values as in eq. (41), (48), and Section C
settings.alpha = {0.05, 0.05, 0.05, 0.05}; % significance level
settings.IV = {[], [], [], []}; % no IVs

% Technical settings (cell arrays are used to loop over the two parameters: theta1 and theta2)
sim = struct;
sim.grid_Theta = {linspace(-40, 100, 1401)', linspace(-40, 100, 1401)'};
sim.rng_seed = 20220826;
sim.num_boots = 1000;
sim.num_robots = 4; % number of parallel workers
sim.sim_name = 'table_3';

results = struct;
results.CI_vec = {zeros(4, 2), zeros(4, 2)}; % specs x {LB, UB}
results.Tn_vec = {zeros(size(sim.grid_Theta{1}, 1), 4), zeros(size(sim.grid_Theta{2}, 1), 4)};
results.comp_time = zeros(4, 1);
results.hat_r_inf = zeros(4, 2); % As in eq. (A.16) in Appendix C

%% 2 Computation
%  three steps:
%             1) compute eq. (4.4) and (4.25) in Andrews and Kwon (2023)
%             2) compute test statistic and critical value
%             3) construct confidence interevals

% delete(gcp)
% parpool('local',sim.num_robots)

for sim0 = 1:4

    disp(['case ' num2str(sim0)])

    tic

    for theta_index = 1:2
        % Temporary in-loop variables (for each theta)
        reject_H = zeros(size(sim.grid_Theta{theta_index}, 1), 1); % save which points reject H0
        Test_vec = zeros(size(sim.grid_Theta{theta_index}, 1), 1);
        cv_vec = zeros(size(sim.grid_Theta{theta_index}, 1), 1);

        % Step 1: compute hat_r_inf and An_vec as in (4.4) and (4.25) in Andrews and Kwon (2023)

        %        An_vec = zeros(1,sim.num_boots);
        %        hat_r_inf = 0;

        if strcmp(settings.test_stat{sim0}, 'RC-CCK') % this step does not depend on the value of theta
            AK_results = compute_An_vec(dgp.D_matrix, dgp.A_matrix, dgp.J0_vec, settings.Vbar{sim0}, settings.IV{sim0}, theta_index, sim.grid_Theta{theta_index}, sim.num_boots, sim.num_robots, sim.rng_seed);
            An_vec = AK_results.An_vec; % input to compute c.value SPUR1, see Section 4.4 in Andrews and Kwon (2023)
            hat_r_inf = AK_results.hat_r_inf; % input to recenter the statistic, see Section 4.1 in Andrews and Kwon (2023)
            results.hat_r_inf(sim0, theta_index) = hat_r_inf;
        end

        % Step 2: find test stat. Tn(theta) and c.value(theta) using G_restriction

        parfor point0 = 1:size(sim.grid_Theta{theta_index}, 1)
            theta0 = [0 0]';
            theta0(theta_index) = sim.grid_Theta{theta_index}(point0, :);

            %test_H0: [T_n, c_value]
            if strcmp(settings.test_stat{sim0}, 'CCK')
                test_H0 = G_restriction(dgp.W_data, dgp.A_matrix, theta0, dgp.J0_vec, settings.Vbar{sim0}, settings.IV{sim0}, theta_index, settings.test_stat{sim0}, settings.cv{sim0}, settings.alpha{sim0}, sim.num_boots, sim.rng_seed);
            elseif strcmp(settings.test_stat{sim0}, 'RC-CCK')
                test_H0 = G_restriction(dgp.W_data, dgp.A_matrix, theta0, dgp.J0_vec, settings.Vbar{sim0}, settings.IV{sim0}, theta_index, settings.test_stat{sim0}, settings.cv{sim0}, settings.alpha{sim0}, sim.num_boots, sim.rng_seed, An_vec, hat_r_inf);
            end

            Test_vec(point0) = test_H0(1);
            cv_vec(point0) = test_H0(2);

            reject_H(point0) = 1 * (test_H0(1) > test_H0(2));
        end

        results.Tn_vec{theta_index}(:, sim0) = Test_vec;

        % Step 3: find confidence intervals using Tn(theta) and c.value(theta)

        % Confidence Interval for thetai

        CS_vec = [];

        for point0 = 1:size(sim.grid_Theta{theta_index}, 1)
            thetai = sim.grid_Theta{theta_index}(point0, :)';

            if reject_H(point0) == 0
                CS_vec = [CS_vec; thetai'];
            end

        end

        if sum(size(CS_vec)) == 0 % it may be the CI is empty
            results.CI_vec{theta_index}(sim0, :) = [NaN NaN];
            [~, point0] = min(Test_vec);
            thetai = sim.grid_Theta{theta_index}(point0, :);
            results.CI_vec{theta_index}(sim0, 2) = thetai; % in this case, we report [nan, argmin test statistic]
        else
            results.CI_vec{theta_index}(sim0, :) = [min(CS_vec) max(CS_vec)];
        end

    end

    toc
    time = toc;
    results.comp_time(sim0, 1) = time;
end

%% 3 Save results
save(fullfile('_results', strcat(sim.sim_name, '.mat')), 'dgp', 'settings', 'sim', 'results');

%% 4 Print table
cd_name = 'tables-tex';
mkdir(fullfile('_results', cd_name));

f = fopen(fullfile('_results', cd_name, strcat(sim.sim_name, '.tex')), 'w'); % Open file for writing

fprintf(f, '%s\n', '\begin{tabular}{c c c c c}');
fprintf(f, '%s\n', '\hline \hline');
fprintf(f, '%s\n', 'Test Stat. & Crit. Value & $\theta_1$: Coca-Cola & $\theta_2$: Energy Brands & Comp. Time \\');
fprintf(f, '%s\n', '\hline');

for row0 = 1:4

    if row0 == 1
        test0 = 'CCK';
    else
        test0 = 'RC-CCK';
    end

    if strcmp(settings.cv{row0}, 'SN2S')
        cvalue0 = 'self-norm';
    elseif strcmp(settings.cv{row0}, 'EB2S')
        cvalue0 = 'bootstrap';
    elseif strcmp(settings.cv{row0}, 'SPUR1')
        cvalue0 = 'SPUR1';
    end

    if strcmp(num2str(results.CI_vec{1}(row0, 1)), 'NaN')

        if strcmp(num2str(results.CI_vec{2}(row0, 1)), 'NaN')
            fprintf(f, '%s%s%s%s%5.1f%s%5.1f%s%5.1f%s\n', test0, ' & ', cvalue0, ' & $', results.CI_vec{1}(row0, 2), '^{\dagger}$ & $', results.CI_vec{2}(row0, 2), '^{\dagger}$ &', results.comp_time(row0, 1), '\\');
        else
            fprintf(f, '%s%s%s%s%5.1f%s%5.1f%s%5.1f%s%5.1f%s\n', test0, ' & ', cvalue0, ' & $', results.CI_vec{1}(row0, 2), '^{\dagger}$ & [', results.CI_vec{2}(row0, 1), ' , ', results.CI_vec{2}(row0, 2), '] &', results.comp_time(row0, 1), '\\');
        end

    else
        fprintf(f, '%s%s%s%s%5.1f%s%5.1f%s%5.1f%s%5.1f%s%5.1f%s\n', test0, ' & ', cvalue0, ' & [', results.CI_vec{1}(row0, 1), ' , ', results.CI_vec{1}(row0, 2), '] & [', results.CI_vec{2}(row0, 1), ' , ', results.CI_vec{2}(row0, 2), '] &', results.comp_time(row0, 1), '\\');

    end

    if row0 == 1 || row0 == 3
        fprintf(f, '%s\n', '\hline');
    end

end

fprintf(f, '%s\n', '\hline \hline');
fprintf(f, '%s', '\end{tabular}');

File: python/table_3.py
""" Table 2 in Section 8.2.1 in Canay, Illanes and Velez (2023)
"""

import time
from pathlib import Path

import numpy as np
import texttable as tt
from joblib import Parallel, delayed
from latextable import draw_latex

import ineq_functions as ineq

# Make folders for results
results_dir = Path(__file__).resolve().parent / "_results"
tables_dir = results_dir / "tables-tex"
tables_dir.mkdir(parents=True, exist_ok=True)


def load_data(name):
    file_path = Path(__file__).resolve().parent.parent / "data" / (name + ".csv")
    return np.loadtxt(file_path, delimiter=",")


dgp = {i: load_data(i) for i in ["A", "D", "J0"]}
dgp["W"] = dgp["D"][:, 1:]
n = dgp["A"].shape[0]

settings = {
    "v_bar": [0, 0, 0, 0],
    "test_stat": ["CCK", "RC-CCK", "RC-CCK", "RC-CCK"],
    "cv": ["SN2S", "SN2S", "EB2S", "SPUR1"],
    "alpha": 0.05,
    "iv": None,
}

sim = {
    "grid_size": 1401,
    "rng_seed": 20220826,
    "bootstrap_replications": 1000,
    "num_robots": 4,
    "sim_name": "table_3",
}
sim["grid_theta"] = (
    np.linspace(-40, 100, sim["grid_size"]),
    np.linspace(-40, 100, sim["grid_size"]),
)

results = {
    "ci_vector": (np.empty((4, 2)), np.empty((4, 2))),
    "tn_vector": (np.empty((sim["grid_size"], 4)), np.empty((sim["grid_size"], 4))),
    "comp_time": np.empty(4),
    "hat_r_inf": np.empty((4, 2)),  # As in eq. (A.16) in Appendix C
}

# Generate bootstrap indices
bootstrap_indices = ineq.helpers.get_bootstrap_indices(
    n, sim["bootstrap_replications"], sim["rng_seed"]
)


# Define the theta0 function
def theta0(theta, the_theta_index):
    """Returns a vector with zeros everywhere except in the theta_index position
    where it returns theta. This is useful in the loop below."""
    the_theta = np.zeros(2)
    the_theta[the_theta_index] = theta
    return the_theta


for sim_i in range(4):
    print("Simulation:", sim_i + 1)
    # Obtain the time at the beginning of the simulation
    tic = time.perf_counter()
    for theta_index in range(2):
        # Step 1: find hat_r_inf and an_vec
        if settings["test_stat"][sim_i] == "RC-CCK" or settings["cv"][sim_i] == "SPUR1":
            # Step 1.1: find hat_r_inf
            rhat_vec = np.array(
                Parallel(n_jobs=sim["num_robots"])(
                    delayed(ineq.andrews_kwon.rhat)(
                        w_data=dgp["W"],
                        a_matrix=dgp["A"],
                        theta=theta0(theta, theta_index),
                        j0_vector=dgp["J0"],
                        v_bar=settings["v_bar"][sim_i],
                        grid0=theta_index + 1,
                        iv_matrix=settings["iv"],
                    )
                    for theta in sim["grid_theta"][theta_index]
                )
            )
            hat_r_inf = np.min(rhat_vec)
            results["hat_r_inf"][sim_i, theta_index] = hat_r_inf
        else:
            hat_r_inf = None

        if settings["cv"][sim_i] == "SPUR1":
            # Step 1.2: find an_vec
            aux1_var = np.array(
                Parallel(n_jobs=sim["num_robots"])(
                    delayed(ineq.andrews_kwon.rhat)(
                        w_data=dgp["W"],
                        a_matrix=dgp["A"],
                        theta=theta0(theta, theta_index),
                        j0_vector=dgp["J0"],
                        v_bar=settings["v_bar"][sim_i],
                        iv_matrix=settings["iv"],
                        grid0=theta_index + 1,
                        adjust=hat_r_inf,
                    )
                    for theta in sim["grid_theta"][theta_index]
                )
            )
            an_vec = ineq.andrews_kwon.compute_an_vec(
                aux1_var,
                hat_r_inf,
                w_data=dgp["W"],
                a_matrix=dgp["A"],
                theta_grid=sim["grid_theta"][theta_index],
                j0_vector=dgp["J0"],
                v_bar=settings["v_bar"][sim_i],
                iv_matrix=settings["iv"],
                grid0=theta_index + 1,
                bootstrap_indices=bootstrap_indices,
            )
        else:
            an_vec = None

        # Step 2: find test stat. Tn(theta) and c.value(theta) using G_restriction
        output = np.array(
            Parallel(n_jobs=sim["num_robots"])(
                delayed(ineq.g_restriction)(
                    theta=theta0(theta, theta_index),
                    w_data=dgp["W"],
                    a_matrix=dgp["A"],
                    j0_vector=dgp["J0"],
                    v_bar=settings["v_bar"][sim_i],
                    alpha=settings["alpha"],
                    grid0=theta_index + 1,
                    iv_matrix=settings["iv"],
                    test0=settings["test_stat"][sim_i],
                    cvalue=settings["cv"][sim_i],
                    bootstrap_indices=bootstrap_indices,
                    an_vec=an_vec,
                    hat_r_inf=hat_r_inf,
                )
                for theta in sim["grid_theta"][theta_index]
            )
        )

        # Because g_function returns a list, and Parallel returns it's runs in
        # a list, test_vec and cv_vec are the columns of the output matrix!
        test_vec = output[:, 0]
        cv_vec = output[:, 1]

        # Theta values for which the null is not rejected
        cs_vec = sim["grid_theta"][theta_index][test_vec <= cv_vec]

        # Create results objects
        results["tn_vector"][theta_index][:, sim_i] = test_vec

        if len(cs_vec) == 0:
            # it may be the CI is empty
            results["ci_vector"][theta_index][sim_i,] = [
                np.NaN,
                sim["grid_theta"][theta_index][test_vec.argmin()],
            ]
        else:
            results["ci_vector"][theta_index][sim_i,] = [cs_vec.min(), cs_vec.max()]

    # Stop the timer
    toc = time.perf_counter()
    results["comp_time"][sim_i] = toc - tic
    print("~> time:", results["comp_time"][sim_i])

# Save results
(results_dir / sim["sim_name"]).mkdir(exist_ok=True)
for key, value in results.items():
    np.save(results_dir / sim["sim_name"] / key, value)

# Make and print the table
tableObj = tt.Texttable(0)

tableObj.set_cols_align(["l", "l", "c", "c", "c"])
tableObj.set_cols_dtype(["i", "t", "t", "t", "f"])

the_table = np.array(settings["v_bar"])
the_table = np.column_stack((the_table, settings["cv"]))
for ci_theta in results["ci_vector"]:
    the_table = np.column_stack(
        (
            the_table,
            [
                "[" + "{:.1f}".format(x[0]) + ", " + "{:.1f}".format(x[1]) + "]"
                for x in ci_theta
            ],
        )
    )
the_table = np.column_stack((the_table, results["comp_time"]))
the_table = np.vstack(
    (
        [
            "$\\Bar{V}$",
            "Crit. Value",
            "$\\theta_1$: Coca-Cola",
            "$\\theta_2$: Energy Brands",
            "Comp. Time",
        ],
        the_table,
    )
)

tableObj.add_rows(the_table)
print(tableObj.draw())

# Save table
with open(tables_dir / (sim["sim_name"] + ".tex"), "w", encoding="utf8") as f:
    f.write(draw_latex(tableObj))
